src/dogroups.c

```diff
@@ -382,14 +401,20 @@ SEXP dogroups(SEXP dt, SEXP dtcols, SEXP groups, SEXP grpcols, SEXP jiscols, SEX
     for (int j=0; j<ngrpcols; ++j) {
       target = VECTOR_ELT(ans,j);
       source = VECTOR_ELT(groups, INTEGER(grpcols)[j]-1);  // target and source the same type by construction above
-      if (SIZEOF(target)==4) {
+      int tsize = SIZEOF(target);
+      if (tsize==4) {
         int *td = INTEGER(target);
         int *sd = INTEGER(source);
         for (int r=0; r<maxn; ++r) td[ansloc+r] = sd[igrp];
-      } else {
+      } else if (tsize==8) {
         double *td = REAL(target);
         double *sd = REAL(source);
         for (int r=0; r<maxn; ++r) td[ansloc+r] = sd[igrp];
+      } else {
+        // #3634 -- CPLXSXP columns have size 16
+        double complex *td = (double complex *)COMPLEX(target);
+        double complex *sd = (double complex *)COMPLEX(source);
+        for (int r=0; r<maxn; ++r) td[ansloc+r] = sd[igrp];
       }
       // Shouldn't need SET_* to age objects here since groups, TO DO revisit.
     }
```

Comment:
There's one more `switch (TYPEOF(target))` just below here (also I don't know what it's doing), I couldn't figure out how to get that one to work though. Attempts:

```
case CPLXSXP : {
  double complex *td = (double complex *)(COMPLEX(target) + thisansloc);
  // attempt 1; fails because double complex doesn't have .r/.i, only CPLXSXP does
  for (int r=0; r<maxn; ++r) { td[r].r = NA_REAL; td[r].i = NA_REAL; }
  // attempt 2; fails because i'm not allowed to assign to RHS of creal()/cimag()
  for (int r=0; r<,maxn; ++r) { creal(td[r]) = NA_REAL; cimag(td[r]) = NA_REAL; }
} break;
```